### PR TITLE
[SPARK-39672][SQL][3.1] Fix removing project before filter with correlated subquery

### PR DESCRIPTION
**Pull request by manuzhang**
### What changes were proposed in this pull request?
Add more checks to`removeProjectBeforeFilter` in `ColumnPruning` and only remove the project if
1. the filter condition contains correlated subquery
2. same attribute exists in both output of child of Project and subquery

### Why are the changes needed?

This is a legitimate self-join query and should not throw exception when de-duplicating attributes in subquery and outer values.

```sql
select * from
(
select v1.a, v1.b, v2.c
from v1
inner join v2
on v1.a=v2.a) t3
where not exists (
  select 1
  from v2
  where t3.a=v2.a and t3.b=v2.b and t3.c=v2.c
)
```

Here's what happens with the current code. The above query is analyzed into following `LogicalPlan` before `ColumnPruning`.
```
Project [a#250, b#251, c#268]
+- Filter NOT exists#272 [(a#250 = a#266) && (b#251 = b#267) && (c#268 = c#268#277)]
   :  +- Project [1 AS 1#273, _1#259 AS a#266, _2#260 AS b#267, _3#261 AS c#268#277]
   :     +- LocalRelation [_1#259, _2#260, _3#261]
   +- Project [a#250, b#251, c#268]
      +- Join Inner, (a#250 = a#266)
         :- Project [a#250, b#251]
         :  +- Project [_1#243 AS a#250, _2#244 AS b#251]
         :     +- LocalRelation [_1#243, _2#244, _3#245]
         +- Project [a#266, c#268]
            +- Project [_1#259 AS a#266, _3#261 AS c#268]
               +- LocalRelation [_1#259, _2#260, _3#261]
```

Then in `ColumnPruning`, the Project before Filter (between Filter and Join) is removed. This changes the `outputSet` of the child of Filter among which the same attribute also exists in the subquery. Later, when `RewritePredicateSubquery` de-duplicates conflicting attributes, it would complain `Found conflicting attributes a#266 in the condition joining outer plan`. 

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add UT.

### CODE REVIEW

**Comment by manuzhang** [comment]
cc @cloud-fan @mgaido91 @viirya

**Comment by cloud-fan** [comment]
how about branches 3.2 and 3.3? also cc @allisonwang-db

**Comment by manuzhang** [comment]
Both of them have https://github.com/apache/spark/pull/30996 so this issue doesn't "exist". Nonetheless, I think it's still a bug in all branches.

**Comment by dongjoon-hyun** [comment]
FYI, cc @viirya and @huaxingao

**Comment by cloud-fan** [comment]
> Later, when RewritePredicateSubquery de-duplicates conflicting attributes, it would complain 

Can we have more detail about it? Maybe we should fix `RewritePredicateSubquery` instead.

**Comment by manuzhang** [comment]
`RewritePredicateSubquery` checks for duplicated attributes in the outputSet of outerPlan and subPlan. In this case, after `ColumnPruning`, the outputSet of outerPlan is `a#250, a#266, b#251, c#268` and that of subPlan is `1#273, a#266, b#267, c#268#277`.
```
     Project [a#250, b#251, c#268]
     +- Filter NOT exists#272 [(a#250 = a#266) && (b#251 = b#267) && (c#268 = c#268#277)]
      :  +- Project [1 AS 1#273, _1#259 AS a#266, _2#260 AS b#267, _3#261 AS c#268#277]
      :     +- LocalRelation [_1#259, _2#260, _3#261]
      +- Join Inner, (a#250 = a#266)
         :- Project [a#250, b#251]
         :  +- Project [_1#243 AS a#250, _2#244 AS b#251]
         :     +- LocalRelation [_1#243, _2#244, _3#245]
         +- Project [a#266, c#268]
            +- Project [_1#259 AS a#266, _3#261 AS c#268]
               +- LocalRelation [_1#259, _2#260, _3#261]
```

Anyway, `removeProjectBeforeFilter` intends to remove `Project` added by `ColumnPruning`, which is the not case here. I'm not sure whether there is a easy way to differentiate generally.

**Comment by cloud-fan** [comment]
> removeProjectBeforeFilter intends to remove Project added by ColumnPruning, which is the not case here.

Which rule added `Project [a#250, b#251, c#268]`?

**Comment by manuzhang** [comment]
It's in the LogicalPlan at the beginning.
```
'Project [*]
+- 'Filter NOT exists#282 []
   :  +- 'Project [unresolvedalias(1, None)]
   :     +- 'Filter ((('t3.a = 'v2.a) AND ('t3.b = 'v2.b)) AND ('t3.c = 'v2.c))
   :        +- 'UnresolvedRelation [v2], [], false
   +- 'SubqueryAlias t3
      +- 'Project ['v1.a, 'v1.b, 'v2.c]
         +- 'Join Inner, ('v1.a = 'v2.a)
            :- 'UnresolvedRelation [v1], [], false
            +- 'UnresolvedRelation [v2], [], false
```

**Comment by cloud-fan** [comment]
ah, if we change `select v1.a, v1.b, v2.c` to `select *`, do we still hit the bug?

**Comment by manuzhang** [comment]
> if we change select v1.a, v1.b, v2.c to select *, do we still hit the bug?

We can't since `t3.a` can't be resolved.

**Comment by cloud-fan** [comment]
OK I understand the issue now. Sometimes a `Project` is useful to avoid name conflicts. However, it should only be a problem in the analysis phase. We may have other rules that remove `Project` if it doesn't change the query schema. I think it's better to fix `RewritePredicateSubquery`, which kinds of re-resolves the column.

**Comment by manuzhang** [comment]
@cloud-fan I've attempted a fix with my understanding of the original change by @mgaido91. However, I'm not able to think of a test case which could throw `AnalysisException`.

**Comment by cloud-fan** [comment]
> This check is not accurate when there's And expression in the Join condition as in this case. Hence, this PR proposes to add a check whether the intersected attributes exist in all the children of the And expression.

Can you explain the rationale? The subquery filter will be turned in to join eventually so it's not very clear to me how to resolve `a#266` in the join condition.

**Comment by cloud-fan** [comment]
OK I think `DeduplicateRelations` needs some fix. Ideally the outer and inner plan should not have conflicting output attributes after analysis, but this local relation + project case is missed.

**Comment by manuzhang** [comment]
`DeduplicateRelations` doesn't exist in 3.1. 

> Can you explain the rationale? 
I'm not sure about it. BTW, could you give an example for why the check is needed in the first place? What if we removed it?

**Comment by cloud-fan** [comment]
After more thoughts, I think we should treat correlated subquery as a join in optimizer rules. So in this case, once we remove the `Project`, the plan becomes invalid, because the subquery's outer reference, which will be in the join condition, becomes ambiguous.

I think your inital approach is the right direction. But let's make it more precise. We should only keep the `Project`, if:
1. the filter condition contains correlated subqueries
2. the subquery's outer references exists in both join sides if we remove the project.

**Comment by manuzhang** [comment]
@cloud-fan Please take another look.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -742,12 +742,18 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
         // We only remove attribute-only project.
-        p2.projectList.forall(_.isInstanceOf[AttributeReference]) =>
+        p2.projectList.forall(_.isInstanceOf[AttributeReference]) &&
+        !hasConflictingAttrsWithSubquery(e, child) =>
       p1.copy(child = f.copy(child = child))
   }
+
+  private def hasConflictingAttrsWithSubquery(e: Expression, child: LogicalPlan): Boolean = {
+    SubqueryExpression.hasInOrCorrelatedExistsSubquery(e) &&
+      child.outputSet.intersect(e.references).nonEmpty
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:755]: To be more precise, we should find out the correlated subqueries and collect their references.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -742,12 +742,28 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
       p1.copy(child = f.copy(child = child))
   }
+
+  private def hasConflictingAttrsWithSubquery(predicate: Expression,
+    child: LogicalPlan): Boolean = {
+    def hasConflictingAttrs(s: SubqueryExpression): Boolean = {
+      s.references.intersect(child.outputSet).nonEmpty
+    }
+
+    splitConjunctivePredicates(predicate).exists { e =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:759]: why do we need `splitConjunctivePredicates`? `predicate.find` should be sufficient.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -742,12 +742,21 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
       if p2.outputSet.subsetOf(child.outputSet) &&
         // We only remove attribute-only project.
-        p2.projectList.forall(_.isInstanceOf[AttributeReference]) =>
+        p2.projectList.forall(_.isInstanceOf[AttributeReference]) &&
+        !hasConflictingAttrsWithSubquery(e, child) =>
       p1.copy(child = f.copy(child = child))
   }
+
+  private def hasConflictingAttrsWithSubquery(predicate: Expression,
+    child: LogicalPlan): Boolean = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:754]:
```suggestion
  private def hasConflictingAttrsWithSubquery(
      predicate: Expression,
      child: LogicalPlan): Boolean = {
```

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -742,12 +742,22 @@ object ColumnPruning extends Rule[LogicalPlan] {
    * order, otherwise lower Projects can be missed.
    */
   private def removeProjectBeforeFilter(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case p1 @ Project(_, f @ Filter(_, p2 @ Project(_, child)))
+    case p1 @ Project(_, f @ Filter(e, p2 @ Project(_, child)))
       if p2.outputSet.subsetOf(child.outputSet) &&
         // We only remove attribute-only project.
-        p2.projectList.forall(_.isInstanceOf[AttributeReference]) =>
+        p2.projectList.forall(_.isInstanceOf[AttributeReference]) &&
+        !hasConflictingAttrsWithSubquery(e, child) =>
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:749]: Can we add a comment here to explain why this check is needed?

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 5]
```diff
@@ -1788,4 +1788,60 @@ class SubquerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
     }
   }
+
+  test("SPARK-39672: Fix removing project before filter with correlated subquery") {
+    withTempView("v1", "v2") {
+      Seq((1, 2, 3), (4, 5, 6)).toDF("a", "b", "c").createTempView("v1")
+      Seq((1, 3, 5), (4, 5, 6)).toDF("a", "b", "c").createTempView("v2")
+
+      def findProjectExec(df: DataFrame): Seq[ProjectExec] = {
+        df.queryExecution.sparkPlan.collect {
```
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:1798]: Would it be better to check the logical plan?

**Review by allisonwang-db — APPROVED** [review]
Thanks for the fix! We should also add it to 3.2/3.3/master to make sure it won't be broken in the future.

**Comment by manuzhang** [comment]
@allisonwang-db thanks for review. Updated.

**Comment by cloud-fan** [comment]
thanks, merging to 3.1/3.2/3.3/master!

**Pull request closed without merging**